django_emarsys/event.py

```diff
@@ -0,0 +1,263 @@
+        log.warning("these Emarsys event names are not "
+                    "configured in settings.EMARSYS_EVENTS: {}"
+                    .format(', '.join(
+                        '"{}"'.format(x.replace('"', '\\"'))
+                            for x in unknown_event_names)))
+
+    if unsynced_event_names:
+        log.warning("these event names in settings.EMARSYS_EVENTS are not "
+                    "known by Emarsys: {}"
+                    .format(', '.join(
+                        '"{}"'.format(x.replace('"', '\\"'))
+                            for x in unsynced_event_names)))
+
+    return num_new_events, num_updated_ids, num_deleted_ids, \
+        unknown_event_names, unsynced_event_names
+
+
+def trigger_automatic_event(event_name, recipient_email, data=None,
+                            create_user_if_needed=True,
+                            contact_data_provider=None):
+    return _trigger_event(event_name=event_name,
+                          recipient_email=recipient_email,
+                          data=data,
+                          create_user_if_needed=create_user_if_needed,
+                          contact_data_provider=contact_data_provider,
+                          source=EventInstance.SOURCE_AUTOMATIC)
+
+
+def trigger_manual_event(event_name, recipient_email, data=None,
+                         create_user_if_needed=True,
+                         contact_data_provider=None):
+    return _trigger_event(event_name=event_name,
+                          recipient_email=recipient_email,
+                          data=data,
+                          create_user_if_needed=create_user_if_needed,
+                          contact_data_provider=contact_data_provider,
+                          source=EventInstance.SOURCE_MANUAL)
+
+
+def get_placeholder_data(event_name, **kwargs):
+    context_provider = get_context_provider(event_name)
+
+    return context_provider(**kwargs)
+
+
+def _trigger_event(event_name, recipient_email, data=None,
+                   create_user_if_needed=True,
+                   contact_data_provider=None,
+                   source=None):
+
+    emarsys_event_id = get_event_id(event_name)
+
+    event = _create_event_instance(
+        event_name=event_name,
+        recipient_email=recipient_email,
+        emarsys_event_id=emarsys_event_id,
+        data=data,
+        source=source)
+
+    if (event.state == EventInstance.STATE_ERROR
+            and event.result_code == '2008'
+            and create_user_if_needed):
+
+        if contact_data_provider is not None:
+            contact_data = contact_data_provider()
+        else:
+            contact_data = {'E-Mail': recipient_email}
+
+        api.create_contact(contact_data)
+        event = _create_event_instance(
+            event_name=event_name,
+            recipient_email=recipient_email,
+            emarsys_event_id=emarsys_event_id,
+            data=data,
+            source=source)
```

Comment:
We'd still trigger the event twice, once with an error message. I think it makes sense to track those `EventInstance`s.